rfcs/0009_Unified_Static_Memory_Planning.md

```diff
@@ -0,0 +1,476 @@
+    --with-parameter-buffer= "name=flash;target=c,accel"
+```
+### Codegen'd Artifacts
+```
+    //Codegen'd artifacts in metadata.c (lib0.c)
+    const TVMModel my_model = {
+       ...
+       .entrypoint = &entrypoint,
+    }
+
+    static int32_t entrypoint(TVMInputs_my_model* inputs, 
+                               TVMOutputs_my_model* outputs, 
+                               TVMContext* context){
+
+         return my_model_main(inputs.input0,
+                              outputs.output0,
+                              context.workspaces.dtcm,
+                              context.workspaces.sram,
+                              context.parameters.itcm,
+                              context.parameters.flash, 
+                              context.resource_handle);
+    }
+```
+```
+// metadata.h
+
+    #define TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE xxxx
+    #define TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE xxxx
+    #define TVM_MY_MODEL_ITCM_PARAMETER_BUFFER_SIZE xxxx
+    #define TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE xxxx
+
+    typedef struct {
+       uint8_t* dtcm;
+       uint8_t* sram;
+    }  TVMWorkspaces_my_model;
+
+    typedef struct {
+       uint8_t* itcm;
+       uint8_t* flash;
+    }  TVMParameters_my_model;
+
+    typedef struct {
+       uint8_t* input0;
+    }  TVMInputs_my_model;
+
+    typedef struct {
+       uint8_t* output0;
+    }  TVMOutputs_my_model;
+```
+### User Application
+```
+    // The User Application 
+        extern  const TVMModel my_model;
+        __attribute__((section( "ITCM" )  const uint8_t   my_model_params_1[TVM_MY_MODEL_ITCM_PARAMETER_BUFFER_SIZE] = <param_1_data>;
+        __attribute__((section( "FLASH" ), aligned( 16 )))  const uint8_t my_model_params_2[TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE] = <param_2_data>;
+        __attribute__((section( "DTCM" )  static uint8_t workspace_buffer_1[TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE];
+        __attribute__((section( "SRAM" ), aligned( 16 )))  static uint8_t workspace_buffer_2[TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE];
+
+    int main(...) {
+         ...
+         TVMContext context;
+         TVMInputs_my_model_1 inputs = {input};
+         TVMOutputs_my_model_1 outputs = {output};
+         TVMWorkspaces_my_model workspaces = {
+             .sram = &workspace_buffer_1,
+             .dtcm = &workspace_buffer_2,
+         };
+         TVMParameters_my_model parameters = {
+             .flash = &my_model_params_1,
+             .itcm = &my_model_params_2
+         };
+         TVMSetWorkspaces(&context, &workspaces);
+         TVMSetParameters(&context, parameters);
+         TVMExecute(&my_model, &inputs, &outputs, &context);
+    }
+```
+# Reference-level explanation
+
+## Overview
+
+This should be a IRModule (TIR) → IRModule (TIR) pass.
+
+Inputs : 
+* AoT TIR PrimFunc ( the control function describing the call graph to operators)
```

Comment:
Hi Chris,

> In the unified lowering flow we may choose to introduce explicit storage and tensor allocations into the relay AST, and then hoist memory planning out of the executors into the unified lowering flow.

> [IRModule(relay)] 
-> transforms (incl. mem. planning) -> [IRModule(relay)]

I dont see much value in doing memory planning at this level as the intra-operator allocate nodes are not visible. Can you elaborate ?

> With customization in the lowering flow, executors such as AoT can allow per target planning, and then generate code by matching allocation nodes in the relay AST and lowering them directly to TIR when building the main function.

Can you explain what do you mean by allocation nodes in the relay AST ? 
In our view, relay is a pure functional language that is designed as an IR to represent operator-level info. It feels wrong to do memory planning at this level.

I guess Im more interested in knowing why graph and AoT, both cannot lower the main function to TIR before performing executor specific lowering ( that is generating the JSON or main function  -- that is the executor specific lowering).

A side point to this is -- as we have discussed in the discuss post -- "I think going forward graph executor might be able to load a packed function of the tvm_main instead of json – it’ll be less confusing as how the graph executor runtime is positioned as of today which is more of a (a very thin – as its supposed to be :slight_smile: ) middleware that connect the graph json and the compiled operator library".

Having said that, I can see this work enabling a path (to extend) towards that – though we only plan to create USMP component that is a TIR IRModule → TIR IRModule which we initially test and support for the AoT executor."

https://discuss.tvm.apache.org/t/rfc-unified-static-memory-planning/10099/2

> with transforms including the ability to customize the transformations applied for different hardware targets. Moreover, transformations are applied at each stage of lowering and one or more of those transforms may include memory planning.

I dont see why this cannot happen if we have the full program in TIR.

> If we push forward with this RFC without co-designing it with unified lower, my concern is that we continue to kick the problem down the road and it will make unifying the lowering and planning for all executors more difficult.

The existence of different executors are for different use-cases, therefore there is only so much unification we could do. IMO, we should only have two executors AoT and JIT (VM). Moreover, I dont feel static memory planning is applicable to the latter. I'd view graph executor as an application that support RPC for AoT that could be used in a tuning process (Refer to the earlier comment on the importance of using a main function instead of JSON).

I'd like to hear why designing a static memory planner that works where the full program is expressed in TIR creates a divergence at least between graph and AoT.

@jroesch @areusch @tqchen @mbaret

Reply:
>In our view, relay is a pure functional language that is designed as an IR to represent operator-level info. It feels wrong to do memory planning at this level.

I believe @jroesch's RFC on unified lowering can help clarify this point some when it lands soon. My take is that overall we would like to move to a world in which the IR is progressively lowered from initially an operator only representation to something closer to hardware, including to TIR for AoT, with target customizable transformation available at every progressively lowered step. 

>I dont see much value in doing memory planning at this level as the intra-operator allocate nodes are not visible. Can you elaborate ?

Partial memory planning can be useful at this level. Let's take the constrained resource (SRAM) scheduling as an example, where SRAM is large enough to hold some full weight tensors. Assume weight pinning in SRAM is a special case of prefetching, and consider scheduling prefetch copy nodes that move weight tensors from DDR to SRAM storage. Depending on how many weights are pinned from the beginning and prefetched at various points in the topologically ordered execution, the amount of SRAM available for intra-operator scratch will change. In unified lowering this SRAM "stack" size could be provided to the TECompiler when doing subgraph scheduling. If this subgraph contains, for example, multiple convolutions and a striping technique is employed, the available SRAM scratch can be used as a constraint when an autoscheduler is searching the schedule space of this complicated subgraph. After scheduling another layer of transformations can occur which can include full / unified memory planning of inter- and intra- op storage.


> The existence of different executors are for different use-cases, therefore there is only so much unification we could do. IMO, we should only have two executors AoT and JIT (VM), conceptually (that is not to say graph executor is not important). 

We can consider a case in which everything is unified behind a MemoryPlan->VM->AoT lowering flow in which the AoT TIR main is generated from the relax VM representation. In this case the resulting AoT TIR also supports dynamism. But even before this we could stage the effort and make early strides towards unifying the lowering flows such that different executors utilize the same common planning infrastructure,
```
MemoryPlan -> VM
          \-> AoT
          \-> Graph
```

>I dont see why this cannot happen if we have the full program in TIR.

I'm not saying that we shouldn't be able to do final memory planning in a full program TIR. Unified lowering should support this. The argument I'm making is that supporting unified static memory planning (USMP) shouldn't be available to AoT only. My feeling is that the approach to USMP should be co-designed as part of the approach to unified lowering so that all executor paths can benefit from the ability to plan inter-op and intra-op storage together.

Reply:
Hi Chris,

> I believe @jroesch's RFC on unified lowering can help clarify this point some when it lands soon. My take is that overall we would like to move to a world in which the IR is progressively lowered from initially an operator only representation to something closer to hardware, including to TIR for AoT, with target customizable transformation available at every progressively lowered step.

Conceptually this is sensible, Im worried we will mix the responsibilities of Relay and TIR, where the former is a functional language and TIR has imperative properties (Stmts) that encapsulate Exprs. I guess my question is why do we say TIR is for AoT ?, In our view, we should not bypass TIR in the lowering, even for the main callgraph lowering that we call as specialized executor lowering. In that world, everything will get represented in TIR (not the just AoT executor).
cc : @tqchen @mbrookhart 

> Partial memory planning can be useful at this level. Let's take the constrained resource (SRAM) scheduling as an example, where SRAM is large enough to hold some full weight tensors. Assume weight pinning in SRAM is a special case of prefetching, and consider scheduling prefetch copy nodes that move weight tensors from DDR to SRAM storage. Depending on how many weights are pinned from the beginning and prefetched at various points in the topologically ordered execution, the amount of SRAM available for intra-operator scratch will change. In unified lowering this SRAM "stack" size could be provided to the TECompiler when doing subgraph scheduling. If this subgraph contains, for example, multiple convolutions and a striping technique is employed, the available SRAM scratch can be used as a constraint when an autoscheduler is searching the schedule space of this complicated subgraph. After scheduling another layer of transformations can occur which can include full / unified memory planning of inter- and intra- op storage.

Isn't this similiar to what could be done using https://tvm.apache.org/docs/api/python/te.html#tvm.te.Schedule.cache_read ? I guess my query is why are we adding Relay AST nodes to do this while we can do this with less changes and holistically in TIR ?

cc : @mbaret 

By doing so we need to duplicate most logic in scheduling primivites and TIR IR nodes to Relay as well. Do we know a strong reason to perform such a duplication ? It is not immediately obvious how this aligns with the goals of the unified lowering.

> We can consider a case in which everything is unified behind a MemoryPlan->VM->AoT lowering flow in which the AoT TIR main is generated from the relax VM representation. In this case the resulting AoT TIR also supports dynamism. But even before this we could stage the effort and make early strides towards unifying the lowering flows such that different executors utilize the same common planning infrastructure,

> MemoryPlan -> VM
          \-> AoT
          \-> Graph

Does supporting dynamism require a relay-level memory plan ?

My argument is unlike lowering it from directly from relay, if we unify the lowering after the whole program is expressed in TIR, it would avoid duplication -- which I think is one of the goals in the unified lowering refactor.

```
Relay --> TECompiler --> TIR (main is also in TIR) --> MemoryPlan --> AoT
                                                                                                           \--> Graph
                                                                                                           \--> VM
```
It'd be interesting to know what relay constructs that are used in VM is not possible to be expressed in TIR. If we can fully express the dynamism in TIR, I'd say we should not bypass the TIR lowering for any executor. If we are not bypassing the TIR lowering for any executor (operators or control code), as stated before, doing a memory plan at relay level seems a bit redundant. Its entirely possible, I might be missing some info on VM, thus feel to enlighten me :).

> Im not saying that we shouldn't be able to do final memory planning in a full program TIR. Unified lowering should support this. The argument I'm making is that supporting unified static memory planning (USMP) shouldn't be available to AoT only. My feeling is that the approach to USMP should be co-designed as part of the approach to unified lowering so that all executor paths can benefit from the ability to plan inter-op and intra-op storage together.

We agree with this fully and additionally we are saying we should not bypass 'full program TIR' for any executor, if possible. All, we are proposing here is Full Program TIR --> Full Program TIR transformation that could be attached to executor codegen pipeline, if each executor have that state (currently the graph bypasses this, it goes from Relay --> JSON directly) in the pipeline (I dont see a strong reason why they should not have that, at least yet). Then, its incremental work to attach it to each executor -- In a similiar way we do TECompiler refactor (attaching the TECompiler for each executor).

cc : @Mousius @mbaret

Reply:
Hi Manupa- we've been having some conversations internally trying to sync up on the unified lowering flow to understand how the USMP RFC and unified lowering can work together. @areusch will follow up with some comments. Overall I think it sounds like we have the same goals and if we can arrive at design in which only incremental refactoring is needed, like with updating AOT to use the TECompiler, then we're good. I'll also follow up directly with some of the questions you raised in response in another post, just didn't want to let the conversation lapse.

Reply:
hi Manupa,

@jroesch @csullivan and I discussed this at length over the last two days in conjunction with @jroesch's plans to unify the Graph, AOT, and VM lowering pipelines. The root of the concern was over three issues:
1. it may be of interest for some users to explicitly specify which memory pool a Relay operator shall place its results. How shall TVM interpret such requests?
2. What if memory planning impacts scheduling (e.g. you may want to choose a different schedule because memory does not fit)?
3. As a side effect of unifying the lowering flows, it will be desirable to also unify memory planning across the three. Given the VM supports additional features over AOT/Graph executors (e.g. dynamic shapes, potentially additional control flow such as recursion), it may not be possible to perform static memory planning on all tensors in such a world. We want to ensure such a unified flow is possible without being overly complex and while still supporting features.
4. There was some confusion as to how GraphExecutor/AOTExecutor may both use USMP.

Point #1 can be handled with a Relay-level annotation on either the function or variable which can then be moved into TIR to drive the memory planner.
Point #4 is handled by writing a pass to translate the USMP-planned AOT top-level function into a GraphExecutor graph.json when possible. I've clarified this with @jroesch and @csullivan.

To point #2: there are a couple of classes of this problem: 1) when a schedule must be discarded because a single buffer always doesn't fit into RAM; 2) when a particular configuration of a schedule can fit into RAM, but not all. Ultimately, this problem can always be addressed by iteratively running the compilation pipeline starting from scheduling and discarding the results after memory planning if they don't work out. However, this solution is very slow so we prefer to avoid it. There are a couple of ways to do this:
1. introduce memory constraints at schedule time so that it's obvious when a buffer is far too big for the available memory
2. We can model allocations before scheduling, like the VM does. this can allow us to discard schedules in class 1 before doing expensive memory planning.
3. we may be able to leverage AutoTIR to "tweak" a schedule (e.g. reduce the intermediate scratchpad usage) when that usage is proportional to an inner/outer loop nest (this solution re

To point #3: After some discussion, it appears that the planning done in the VM and USMP have something in common: they both create allocate nodes to model the graph-level tensors in TIR. Since USMP relies on building an abstract representation of allocate/free events, it should be possible to continue to build this after unifying the flows based on either the AOT top-level function or a similar function produced by lowering the VM-optimized Relay module to TIR. Therefore, the static tensors used in a TIR IRModule should still be able to be planned using USMP after unifying the flows. The dynamic tensors may need to be handled separately, but in general doing much ahead-of-time planning for tensors with data-dependent lifetime (e.g. when handling recursion) is difficult.

Finally, related: @jroesch also mentioned that in unifying the flows, he wants to move to the world where a conceptual equivalent of the VM buffer/storage nodes are injected pre-scheduling. There will be some work in doing this to ensure that they are fully compatible with the memory pool concept here, but at first glance they appear to be broadly compatible.  The VM carries a somewhat similar concept of a per-device dynamic allocator, which is conceptually similar to that which may be required if we were to consider at a later time supporting dynamic models in the C runtime. We think these concepts are similar and can be unified moving forwards. 

Here our intent isn't to choose any one implementation over another but purely to ensure we can reduce the conceptual complexity of the compiler over time while retaining flexibility of the compiler. My thoughts are that adopting this approach (and therefore, going ahead with USMP now) should allow others to begin building the static memory planning algorithms while we are working to simplify the compiler, and this will ensure we can still continue to support any memory planning algorithms developed on top of USMP in the medium-long term.

Please let me know if this makes sense to you! I think this should allow everyone to continue working without stepping on each others' toes excessively.

-Andrew

Reply:
Hi @areusch @csullivan,

Thanks for the clarification and it makes sense for us.